client/auth.go

```diff
@@ -69,35 +77,53 @@ func (c *Conn) readInitialHandshake() error {
 	pos += 2
 
 	if len(data) > pos {
-		// skip server charset
+		// default server a_protocol_character_set, only the lower 8-bits
 		// c.charset = data[pos]
 		pos += 1
 
 		c.status = binary.LittleEndian.Uint16(data[pos : pos+2])
 		pos += 2
-		// capability flags (upper 2 bytes)
+
+		// The upper 2 bytes of the Capabilities Flags
 		c.capability = uint32(binary.LittleEndian.Uint16(data[pos:pos+2]))<<16 | c.capability
 		pos += 2
 
-		// auth_data is end with 0x00, min data length is 13 + 8 = 21
-		// ref to https://dev.mysql.com/doc/internals/en/connection-phase-packets.html#packet-Protocol::Handshake
-		maxAuthDataLen := 21
-		if c.capability&CLIENT_PLUGIN_AUTH != 0 && int(data[pos]) > maxAuthDataLen {
-			maxAuthDataLen = int(data[pos])
+		// length of the combined auth_plugin_data (scramble), if auth_plugin_data_len is > 0
+		authPluginDataLen := data[pos]
+		if (c.capability&CLIENT_PLUGIN_AUTH == 0) && (authPluginDataLen > 0) {
+			return errors.Errorf("invalid auth plugin data filler %d", authPluginDataLen)
 		}
+		pos++
 
-		// skip reserved (all [00])
-		pos += 10 + 1
+		// skip reserved (all [00] ?)
+		pos += 10
 
-		// auth_data is end with 0x00, so we need to trim 0x00
-		resetOfAuthDataEndPos := pos + maxAuthDataLen - 8 - 1
-		c.salt = append(c.salt, data[pos:resetOfAuthDataEndPos]...)
+		if c.capability&CLIENT_SECURE_CONNECTION != 0 {
+			// Rest of the plugin provided data (scramble)
 
-		// skip reset of end pos
-		pos = resetOfAuthDataEndPos + 1
+			// $len=MAX(13, length of auth-plugin-data - 8)
+			rest := int(authPluginDataLen) - 8
+			if max := 13; rest > max {
+				rest = max
+			}
+			if data[pos+rest-1] != 0 {
```

Comment:
```suggestion
			if data[pos+rest] != 0 {
```

BTW, im not familiar with this part so double check that why there's NULL after scramble? In the MySQL protocol doc I think it's a fixed length string, rather than a NULL-terminated string.

Reply:
I found [the origin](https://github.com/mysql/mysql-server/blob/c47fab04/sql/auth/sql_authentication.cc#L584-L585) of it:
>n - rest of the plugin provided data (at least 12 bytes)
>1 - \0 byte, terminating the second part of a scramble

And [sha256_password auth implementation](https://github.com/mysql/mysql-server/blob/a153bf585e827818641fb534296be3e9a44db466/sql/auth/sql_authentication.cc#L4767-L4771) follows this rule: 
>Native authentication sent 20 bytes + '\0' character = 21 bytes.
>This plugin must do the same to stay consistent with historical behavior

Also, alibaba channel [contains the same description](https://github.com/alibaba/canal/blob/0ec46991499a22870dde4ae736b2586cbcbfea94/driver/src/main/java/com/alibaba/otter/canal/parse/driver/mysql/packets/server/HandshakeInitializationPacket.java#L94-L95):
>Packet规定最后13个byte是剩下的scrumble, 但实际上最后一个字节是0, 不应该包含在scrumble中.

*I used a translator for this.*


I can remove this check by \0, but it looks like a historic standard convention.

Reply:
What's about 20 bytes len I found [this](https://github.com/mysql/mysql-server/blob/1bfe02bdad6604d54913c62614bde57a055c8332/sql/auth/sql_authentication.cc#L1641-L1642):
>the first packet *must* have at least 20 bytes of a scramble.
>if a plugin provided less, we pad it to 20 with zeros

Right now I don't see places where it would be more than 20 bytes, but let it be for the future.

Reply:
I checked this logic for mysql 8.0 and 5.7. Unfortunately, compatibility with 5.6- was broken since we added `json` fields in our tests in 2021.